crates/fj-kernel/src/validate/cycle.rs

```diff
@@ -1,12 +1,132 @@
 use crate::objects::Cycle;
+use crate::objects::HalfEdge;
+use fj_math::Point;
+use fj_math::Scalar;
+use itertools::Itertools;
 
 use super::{Validate, ValidationConfig, ValidationError};
 
 impl Validate for Cycle {
     fn validate_with_config(
         &self,
-        _: &ValidationConfig,
-        _: &mut Vec<ValidationError>,
+        config: &ValidationConfig,
+        errors: &mut Vec<ValidationError>,
     ) {
+        CycleValidationError::check_half_edges_disconnected(
+            self, config, errors,
+        )
+    }
+}
+
+/// [`Cycle`] validation failed
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum CycleValidationError {
+    /// [`Cycle`]'s half-edges are not connected
+    #[error(
+        "Adjacent `HalfEdge`s are distinct\n\
+        - End position of first `HalfEdge`: {end_of_first:?}\n\
+        - Start position of second `HalfEdge`: {start_of_second:?}\n\
+        - `HalfEdge`s: {half_edges:#?}"
+    )]
+    HalfEdgesDisconnected {
+        /// The end position of the first [`HalfEdge`]
+        end_of_first: Point<2>,
+
+        /// The start position of the second [`HalfEdge`]
+        start_of_second: Point<2>,
+
+        /// The distance between the two vertices
+        distance: Scalar,
+
+        /// The half-edge
+        half_edges: Box<(HalfEdge, HalfEdge)>,
+    },
+    #[error("Expected at least one `HalfEdge`\n")]
+    NotEnoughHalfEdges,
+}
+
+impl CycleValidationError {
+    fn check_half_edges_disconnected(
+        cycle: &Cycle,
+        config: &ValidationConfig,
+        errors: &mut Vec<ValidationError>,
+    ) {
+        // If there are no half edges
+        if cycle.half_edges().next().is_none() {
+            errors.push(Self::NotEnoughHalfEdges.into());
+            return;
+        }
+        for (first, second) in cycle
+            .half_edges()
+            // Chain the first half_edge so that we make sure that the last connects to the first.
+            // This unwrap will never fail because we checked before that there are enough half_edges.
+            .chain(std::iter::once(cycle.half_edges().next().unwrap()))
+            .tuple_windows()
```

Comment:
Also fine as-is, but there's an easier (and less panicky) way to do the same thing:
```suggestion
            .circular_tuple_windows()
```